node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/AthenzServiceIdentityHttpClient.java

```diff
@@ -0,0 +1,90 @@
+ * and that uses {@link SSLContext} provided by {@link SiaIdentityProvider}.
+ *
+ * This client also supports live updates of credentials by implementing
+ * {@link ServiceIdentityProvider.Listener}.
+ *
+ * @author freva
+ */
+class AthenzServiceIdentityHttpClient implements AutoCloseable, ServiceIdentityProvider.Listener {
+    private final Map<String, ConnectionSocketFactory> socketFactoryRegistry = new ConcurrentHashMap<>();
+    private final AthenzIdentityVerifier identityVerifier;
+    private final CloseableHttpClient httpClient;
+
+    AthenzServiceIdentityHttpClient(Set<AthenzIdentity> allowedIdentities) {
+        DefaultHttpClientConnectionOperator httpClientConnectionOperator =
+                new DefaultHttpClientConnectionOperator(socketFactoryRegistry::get, null, null);
+        PoolingHttpClientConnectionManager cm =
+                new PoolingHttpClientConnectionManager(httpClientConnectionOperator, null, -1, TimeUnit.MILLISECONDS);
+        cm.setMaxTotal(200); // Increase max total connections to 200, which should be enough
+
+        // Have experienced hang in socket read, which may have been because of
+        // system defaults, therefore set explicit timeouts. Set arbitrarily to
+        // 15s > 10s used by Orchestrator lock timeout.
+        int timeoutMs = 15_000;
+        RequestConfig requestBuilder = RequestConfig.custom()
+                .setConnectTimeout(timeoutMs) // establishment of connection
+                .setConnectionRequestTimeout(timeoutMs) // connection from connection manager
+                .setSocketTimeout(timeoutMs) // waiting for data
+                .build();
+
+        this.socketFactoryRegistry.put("http", PlainConnectionSocketFactory.getSocketFactory());
+        this.identityVerifier = new AthenzIdentityVerifier(allowedIdentities);
+        this.httpClient = HttpClientBuilder.create()
+                .setDefaultRequestConfig(requestBuilder)
+                .disableAutomaticRetries()
+                .setUserAgent("node-admin")
+                .setConnectionManager(cm).build();
+    }
+
+    AthenzServiceIdentityHttpClient(SiaIdentityProvider identityProvider, Set<AthenzIdentity> allowedIdentities) {
+        this(allowedIdentities);
+        onCredentialsUpdate(identityProvider.getIdentitySslContext(), null);
+    }
+
+    public CloseableHttpResponse execute(HttpUriRequest request) throws IOException {
+        return httpClient.execute(request);
+    }
+
+    @Override
+    public void close() throws IOException {
+        httpClient.close();
+    }
+
+    @Override
+    public void onCredentialsUpdate(SSLContext sslContext, AthenzService identity) {
+        SSLConnectionSocketFactory sslConnectionSocketFactory = new SSLConnectionSocketFactory(sslContext, identityVerifier);
+        setSslConnectionSocketFactory(sslConnectionSocketFactory);
+    }
+
+    private void setSslConnectionSocketFactory(SSLConnectionSocketFactory sslConnectionSocketFactory) {
```

Comment:
Of course the cleanest approach :)

Reply:
Just need to find a way to test this... :thinking: